packages/runtime/datastore/src/channelDeltaConnection.ts

```diff
@@ -55,17 +92,41 @@ export class ChannelDeltaConnection implements IDeltaConnection {
 	}
 
 	public reSubmit(content: any, localOpMetadata: unknown) {
-		this.handler.reSubmit(content, localOpMetadata);
+		process(content, localOpMetadata, this.handler.reSubmit.bind(this.handler));
 	}
 
 	public rollback(content: any, localOpMetadata: unknown) {
 		if (this.handler.rollback === undefined) {
 			throw new Error("Handler doesn't support rollback");
 		}
-		this.handler.rollback(content, localOpMetadata);
+		process(content, localOpMetadata, this.handler.rollback.bind(this.handler));
 	}
 
 	public applyStashedOp(content: any): unknown {
-		return this.handler.applyStashedOp(content);
+		try {
+			this.statedOpMd = createStashedOpMetadata();
+			const md = this.handler.applyStashedOp(content);
+			if (md !== undefined) {
+				// temporary assert while we migrate dds to the new pattern
+				// after migration we will always return statedOpMd
+				assert(
+					(this.statedOpMd?.length ?? 0) === 0,
+					"statedOpMd should be empty if metadata returned. this means ops were submitted and local op metadata was returned. this will cause a corruption",
+				);
+				return md;
+			}
+			return this.statedOpMd;
+		} finally {
+			this.statedOpMd = undefined;
+		}
+	}
+
+	private statedOpMd: StashedObjectMetadata | undefined;
```

Comment:
not really clear to me why you chose this name

Reply:
stated is a typo and i don't really like StashedObjectMetadata. been doing lots of porting between prototype branches, and just missed making this better.